modules/nf-core/bwa/mem/main.nf

```diff
@@ -8,13 +8,17 @@ process BWA_MEM {
         'biocontainers/mulled-v2-fe8faa35dbf6dc65a0f7f5d4ea12e31a79f73e40:a34558545ae1413d94bde4578787ebef08027945-0' }"
 
     input:
-    tuple val(meta), path(reads)
+    tuple val(meta) , path(reads)
     tuple val(meta2), path(index)
+    tuple val(meta3), path(fasta)
     val   sort_bam
 
     output:
-    tuple val(meta), path("*.bam"), emit: bam
-    path  "versions.yml"          , emit: versions
+    tuple val(meta), path("*.bam")  , emit: bam,    optional: true
```

Comment:
Sorry, to be really annoying and add another comment, does it make sense to have channels for output bam AND cram? Only one or the other is produced, right?

Could we not just have one non-optionoal 'alignment' output, and one 'index', globbing for multiple extensions? Or will that create difficulties?

Reply:
It's against the guidelines 😄 
I kind of like having more granular control over which outputs I'm using TBH, that way I can fail faster if something goes wrong with the module config.

e.g, the output will always be bam, unless I configure it otherwise. With the global glob I'll always get some form of output, but not necessarily the one I want.

Reply:
https://nf-co.re/docs/contributing/modules#inputoutput-options

```
3. Named file extensions MUST be emitted for ALL output channels e.g. path "*.txt", emit: txt.
```

Reply:
Ahh, fair enuf

Reply:
Still, this means that if I were to switch alignment format in config I'd have to rewire the workflow.

Reply:
Yes it does

Reply:
My view is that the rule is there to prevent the `*` pattern. In any case it doesn't forbid bundling multiple file extensions.
It's not just samtools, I see a lot of tools transparently support BAM/CRAM, or fasta/fastq/fasta.gz/fastq.gz, etc, and we generally don't list every possible file extension as its own channel.

Reply:
Will the `optional` blow up on us? What does the error look like if there's no bam?

I'm wondering if we could do something like 

```
  input:
    tuple val(meta) , path(reads)
    tuple val(meta2), path(index)
    tuple val(meta3), path(fasta)
    val   sort_bam
    val   output_type

    output:
    tuple val(meta), path("*.${output_type}")  , emit: aligned_reads
    tuple val(meta), path("*.${fancy logic to set bai/csi") , emit: index
    path  "versions.yml"            , emit: versions
```

Reply:
How do you mean blow up? If there's no bam or cram it means the tool failed and we should get an exit code corresponding to that.

I've made modules in the past with the fancy logic you suggest, but I got pulled back because it was against the guidelines 🤷🏻